v2/spanner-change-streams-to-sharded-file-sink/src/test/java/com/google/cloud/teleport/v2/templates/SpannerChangeStreamToGcsCustomShardIT.java

```diff
@@ -0,0 +1,296 @@
+   *
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  @Before
+  public void setUp() throws IOException, InterruptedException {
+    skipBaseCleanup = true;
+    synchronized (SpannerChangeStreamToGcsSimpleIT.class) {
+      testInstances.add(this);
+      if (jobInfo == null) {
+        createGcsResourceManager();
+        createSpannerDatabase();
+        createAndUploadShardConfigToGcs();
+        uploadSessionFileToGcs();
+        createSpannerMetadataDatabase();
+        createAndUploadJarToGcs();
+        launchReaderDataflowJob();
+      }
+    }
+  }
+
+  @Test
+  public void testMultiShardsRecordWrittenToGcsWithCustomShardId()
+      throws IOException, java.lang.InterruptedException {
+    // Construct a ChainedConditionCheck with below stages.
+    // 1. Wait for the metadata table to have the start time of reader job
+    // 2. Write 2 records per shard to Spanner
+    // 3. Wait on GCS to have the files
+    // 4. Match the PK in GCS with the PK written to Spanner
+    ChainedConditionCheck conditionCheck =
+        ChainedConditionCheck.builder(
+                List.of(
+                    SpannerRowsCheck.builder(
+                            spannerMetadataResourceManager, "spanner_to_gcs_metadata")
+                        .setMinRows(1)
+                        .setMaxRows(1)
+                        .build()))
+            .build();
+    // Wait for conditions
+    PipelineOperator.Result result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo, Duration.ofMinutes(10)), conditionCheck);
+    // Assert Conditions
+    assertThatResult(result).meetsConditions();
+    // Perform writes to Spanner
+    writeSpannerDataForSingers(1, "one", "");
+    writeSpannerDataForSingers(2, "two", "");
+    writeSpannerDataForSingers(3, "three", "");
+    writeSpannerDataForSingers(4, "four", "");
+
+    // Assert file present in GCS with the needed data
+    assertFileContentsInGCSForMultipleShards();
+  }
+
+  private void writeSpannerDataForSingers(int singerId, String firstName, String shardId) {
+    // Write a single record to Spanner
+    Mutation m =
+        Mutation.newInsertOrUpdateBuilder("Singers")
+            .set("SingerId")
+            .to(singerId)
+            .set("FirstName")
+            .to(firstName)
+            .set("migration_shard_id")
+            .to(shardId)
+            .build();
+    spannerResourceManager.write(m);
+  }
+
+  private void assertFileContentsInGCSForMultipleShards() throws java.lang.InterruptedException {
+    List<Artifact> artifactsShardA = null;
+    List<Artifact> artifactsShardB = null;
+    Thread.sleep(
+        180000); // wait sufficiently for the file to be generated. It takes about 3 minutes
+    // at-least. If not present wait additional 3 minutes before failing
+    for (int i = 0; i < 10; i++) {
+      Thread.sleep(18000); // wait for total 3 minutes over an interval of 18 seconds
```

Comment:
Should this be modeled as a `conditionCheck`? Maybe something like `GCSRowCheck`?

https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/c98fed7d7bd7cd2a8a2ac51c97be1f0706d79178/it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/spanner/conditions/SpannerRowsCheck.java#L27

@darshan-sj 

Seems like the goal here is to wait for a condition to become true, and we should keep checking if the condition becomes true until the timeout limit, which is what we do in SpannerRowCheck.

Reply:
Makes sense. Added